### ISSUE TITLE
Add native theme support (& dark mode) to the message display

### ISSUE THREAD

**Issue by Standard8** [thunderbird-conversations/thunderbird-conversations#1611]
We should do this after #1608 and #1610 since we'll probably have learnt a few things from those.

Hopefully we can do this in a similar way, though I suspect there will be a few more layout issues to resolve.

**Comment by ryanleesipes** [comment]
Hey, how is this coming along? This would be a great improvement!

**Comment by loyukfai** [comment]
Just wondering, if the author of Dark Reader is willing, is it possible for him/her/they to adapt DR to work with your add-on?

Cause it may save time and no need to re-invent the wheel...

Cheers.

**Comment by ariccb** [comment]
would love this to work with dark mode!

**Comment by Juraj-Masiar** [comment]
Hello everyone,
I'm the author of [Darko](https://addons.thunderbird.net/en-US/thunderbird/addon/darko_t/) addon for Thunderbird and I can give you "simple" solution for the dark mode that my addon is using.
It's not as good as the "Dark reader" but it's trivial to implement because it's just a single short CSS with only two styles.
First one will invert the page and shift the hue to fix the colors.
Second will revert this operation on images.
The end!
```CSS
@media not print {
    body:not(.no-darko) {
        filter: invert(100%) hue-rotate(180deg) !important;
        background: #2A2A2E !important;
    }
    body:not(.no-darko) img {
        filter: invert(100%) hue-rotate(-180deg) !important;
    }
}
```
Plus it's disabled for printing using media query and it can be manually disabled by adding ".no-darko" class to the body.

But again, it's not perfect solution because it doesn't detect already dark emails and the text clarity may be slightly reduced. Still, in the evening it's for sure 100% better than looking at the white background so I love it :).

**Comment by martinderm** [comment]
Any news on the issue?

**Comment by cyb456** [comment]
hope this gets added soon

**Comment by godamanx** [comment]
This is how bad it is. I want to change my email client but conversations important for me. But also dark mode too.

![Ekran görüntüsü 2023-11-01 121427](https://github.com/thunderbird-conversations/thunderbird-conversations/assets/4149142/c81e55bc-2796-4965-bc48-93bbcc5394a0)

**Comment by vysmaty** [comment]
Based on @Juraj-Masiar's post, I added the following as a workaround to _conversation.css_ for now:

```css
@media screen and (prefers-color-scheme: dark) {
  .messageBody iframe:not(.no-darko) {
      filter: invert(89%) hue-rotate(180deg) !important;
      background: #27272a !important;
  }
  .messageBody iframe:not(.no-darko) img {
      filter: invert(89%) hue-rotate(-180deg) !important;
  }
}
```

It's not pretty. I can't target the iframe directly. I don't know how. Better than a white wire in the eye. 😇

**Comment by Razer0123** [comment]
Waiting for this too, i should not have to decide between showing sent messages and dark mode

**Comment by vysmaty** [comment]
> Waiting for this too, i should not have to decide between showing sent messages and dark mode

Yeah, that would be great. I've gotten used to it in color inversion. Really fast.

![image](https://github.com/thunderbird-conversations/thunderbird-conversations/assets/10959584/13854a10-f6b5-4ff0-9309-f3502ba8ce17)

_I love this extension. Thank you for it._

**Comment by Razer0123** [comment]
> > Waiting for this too, i should not have to decide between showing sent messages and dark mode
> 
> Yeah, that would be great. I've gotten used to it in color inversion. Really fast.
> 
> ![image](https://user-images.githubusercontent.com/10959584/284953565-13854a10-f6b5-4ff0-9309-f3502ba8ce17.png)
> 
> _I love this extension. Thank you for it._

How did you do this?

**Comment by Razer0123** [comment]
How did you do that?Il giorno 22 nov 2023, alle ore 16:30, martin ***@***.***> ha scritto:﻿

Waiting for this too, i should not have to decide between showing sent messages and dark mode

Yeah, that would be great. I've gotten used to it in color inversion. Really fast.

I love this extension. Thank you for it.

**Comment by im007** [comment]
+1 I tried the custom CSS above, but it's not working for me.
Having a dark mode option in the add-on settings would be amazing.

**Comment by vysmaty** [comment]
> +1 I tried the custom CSS above, but it's not working for me. Having a dark mode option in the add-on settings would be amazing.

Yeah, something's changed. The author knows why he doesn't do hacks like this. 😉

**Comment by Juraj-Masiar** [comment]
Hello, author of Darko addon here again.
I'm getting requests making integration with this addon, which is currently not possible (blocked by: https://github.com/thunderbird-conversations/thunderbird-conversations/issues/1609).

However, I still believe it should be simple enough to implement in this "natively" in this addon (if you are able to inject CSS and JS into the email content in the iframe).

This is the high level architecture view for a "simple dark mode":
1. make sure this CSS is injected into each email body (so each "iframe"), I would say even simple `content_scripts` entry in the Manifest could work:
```css
@media not print {
    body.tc_dark_mode {
        filter: invert(100%) hue-rotate(180deg) !important;
        background: #2A2A2E !important;
    }
    body.tc_dark_mode :is(img, [style*="background-image:"]:not([style*="background-image: none"]), [background*="."], g-emoji) {
        filter: invert(100%) hue-rotate(180deg) !important;
    }
}
```
As you can see, the style is used only if the `body` has the `.tc_dark_mode` class set. So this alone will not change existing behavior!

2. at this point, all you need to do to activate the dark mode, is to add `.tc_dark_mode` class to the body element.
For example, executing this should activate it in all iframes in a specific tab:
```js
await browser.tabs.executeScript(tabId, {
    code: `document.body.classList.toggle('tc_dark_mode', true)`,
    runAt: 'document_end',
    allFrames: true,
  })
```
3. add a button somewhere to UI that executes the code above - that will be the "toggle" button.
4. since we want the dark mode to stay active, the code above should also auto-execute for each opened message, if the "toggle" state is "on".

Although I have a feeling `executeScript` will work only in normal tabs, not in the main Thunderbird tab, so for those cases you would have execute content script into all mails and then use messaging to let them know what to do.

Just an idea... I know we are all very busy here.

**Comment by philia897** [comment]
Hi guys,

These days I'm also thinking about changing to dark mode, and do not want to give up this excellent addon. Today I finally find some time to do it.

Appreciate the solution from @Juraj-Masiar, I made a new pull request to add this feature. Though I did not add any button, I add one more option in the add-on options to enable dark mode.

Here is how it looks like:

![dark-mode](https://github.com/user-attachments/assets/bf10e221-fe1b-4f5e-bc29-3af6ca302555)
![options](https://github.com/user-attachments/assets/fb7e2c84-b673-4ecf-98f2-43f9c54debd6)

**Comment by jceb** [comment]
I also think the padding should be increased. As pointed out in #2149 the text starts right at the border. In dark mode without the dark message background, the experience is not great. Seeing the latest screenshot with a dark message background could also use some padding in my opinion.

![image](https://github.com/user-attachments/assets/ab92d7a8-3265-48d1-a380-aa76e338990e)

**Comment by ajayd-san** [comment]
Hey, What's the status on this? 

I can help writing this feature if it needs help.

**Comment by Juraj-Masiar** [comment]
Note that Thunderbird is getting proper dark mode (even in messages) soon:
https://blog.thunderbird.net/2025/03/thunderbird-release-channel-update/
I'm not sure though how will it affect this addon.

**Comment by vt116** [comment]
> > Note that Thunderbird is getting proper dark mode (even in messages) soon: [blog.thunderbird.net/2025/03/thunderbird-release-channel-update](https://blog.thunderbird.net/2025/03/thunderbird-release-channel-update/) I'm not sure though how will it affect this addon.
> 
> It already have it with 136 (I got to use it while the addon was broken with TB 136), and the result is pretty nice.



> Note that Thunderbird is getting proper dark mode (even in messages) soon: https://blog.thunderbird.net/2025/03/thunderbird-release-channel-update/ I'm not sure though how will it affect this addon.

Unfortunately this doesn't affect the app's inability to do dark mode.

**Comment by Standard8** [comment]
> Is the toggle button in the message header toolbar supposed to revert to "Dark Mode on" on every new message/thread selection or should the state remain?

It is per-message being displayed. The state is not saved. Generally I'm assuming that wanting to switch to light mode is temporary if there's some specific issue in an individual message. Additionally Thunderbird doesn't currently offer a reliable way of storing per-message states.

> Also a question about the toggle button, is it intentional it has some underline around it making it a little bit out of place compared to the other buttons without it on the left and right?

It works fine on Mac.


Reminder: As I mentioned before, please file new issues for bugs, otherwise they won't be tracked.

**Comment by Standard8** [comment]
> Right, so in this version (thunderbird 128esr), your new update makes TB Conversations appear dark with a light theme, and light in the following way with dark theme:

Reminder: As I mentioned before, please file new issues for bugs, otherwise they won't be tracked.

**Comment by vt116** [comment]
> > Right, so in this version (thunderbird 128esr), your new update makes TB Conversations appear dark with a light theme, and light in the following way with dark theme:
> 
> Reminder: As I mentioned before, please file new issues for bugs, otherwise they won't be tracked. This also works for me.

Ok sure, please find the issue link here: https://github.com/thunderbird-conversations/thunderbird-conversations/issues/2212
